### PR TITLE
ENH doc is now built with the new sphinx build

### LINKED ISSUE #5798: Use Makefile for sphinx build
#5743 reminded me:

We still use an ancient `make.py` that is based on what early versions of `sphinx-quickstart` used to generate.  We should probably modernize this to use a Makefile like current versions of sphinx create.

This is sort of low-priority, because the only _real_ advantage would be consistency with other Python projects.  The Makefile won't magically make things faster or anything.  Also, the work is non-trivial since our `make.py` includes customizations and other things that get run that will need to be ported to the "current way to do things" (probably Sphinx extensions).

### PR DESCRIPTION
**Pull request by NelleV**
Instead of relying on make.py which is the antiquated way of building sphinx
documentation, the build now relies on sphinx-build through a makefile (or
make.bat for windows).

- [x] Update the documentation.
- [x] Update travis's build system to reflect this change.

closes #5798

### CODE REVIEW

**Comment by NelleV** [comment]
Unlike before, the -W option is not set. This option makes the build fail with sphinx 1.4.1 with the following error:
```
Warning, treated as error:
WARNING: while setting up extension ['?', 'matplotlib.sphinxext.mathmpl']: node class 'latex_math' is already registered, its visitors will be overridden
```

**Comment by jenshnielsen** [comment]
👍 but this will need an update to travis.yml to call make.py and an option to build the documents with warnings as errors on Travis so we can ensure that warnings don't creep back into the documentation.

File: doc/Makefile [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+	@echo
+	@echo "Build finished. The XML files are in $(BUILDDIR)/xml."
+
+pseudoxml:
+	$(SPHINXBUILD) -b pseudoxml $(ALLSPHINXOPTS) $(BUILDDIR)/pseudoxml
+	@echo
+	@echo "Build finished. The pseudo-XML files are in $(BUILDDIR)/pseudoxml."
+
+
+prepare: mpl_examples mpl_toolkits/axes_grid1/examples \
```
Comment (QuLogic) [inline doc/Makefile]: Could we not just add these symlinks in git?

Comment (tacaswell) [inline doc/Makefile]: symlinks + git + windows do not always play nice

File: doc/README.txt [hunk 2]
```diff
@@ -33,11 +33,9 @@ python documentation system built on top of ReST.  This directory contains
 * mpl_examples - a link to the matplotlib examples in case any
   documentation wants to literal include them
 
-To build the HTML documentation, install sphinx (1.0 or greater
-required), then type "python make.py html" in this directory.  Wait
-for the initial run (which builds the example gallery) to be done,
-then run "python make.py html" again. The top file of the results will
-be ./build/html/index.html
+To build the HTML documentation, install sphinx (1.0 or greater required),
+then type "make html" in this directory .The top file of the results will be
```
Comment (QuLogic) [inline doc/README.txt]: The period and space are twiddled.

File: doc/Makefile [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+# Internal variables.
+PAPEROPT_a4     = -D latex_paper_size=a4
+PAPEROPT_letter = -D latex_paper_size=letter
+ALLSPHINXOPTS   = -d $(BUILDDIR)/doctrees $(PAPEROPT_$(PAPER)) $(SPHINXOPTS) .
+# the i18n builder cannot share the environment and doctrees with the others
+I18NSPHINXOPTS  = $(PAPEROPT_$(PAPER)) $(SPHINXOPTS) .
+
+.PHONY: help clean html dirhtml singlehtml pickle json htmlhelp qthelp devhelp epub latex latexpdf text man changes linkcheck doctest gettext
+
+all: prepare html
```
Comment (QuLogic) [inline doc/Makefile]: The `prepare` shouldn't be here; don't any of the other targets require the same preparation?

Comment (NelleV) [inline doc/Makefile]: I am still testing this. Some targets have this action as dependency.

File: doc/Makefile (hunk 1 above)
Comment (NelleV) [inline doc/Makefile]: Actually, this is what I did. It is much more robust than relying on python code or makefiles to them.
I don't remember ever having a problem with symlinks and windows: is that a recent issue?

Another problem if we do not rely on symlinks is that we'll have to change the make.bat, which is the window's makefile. I am not familiar with those.

**Comment by tacaswell** [comment]
And admittedly old link to the symlink issue is https://stackoverflow.com/questions/5917249/git-symlinks-in-windows

@janschulz @jbmohler Can either of you confirm that the docs can still be built in windows?

**Comment by NelleV** [comment]
Should I remove the symlinks and reintegrate automatically symlinking/copying those in the makefile?

**Comment by tacaswell** [comment]
I agree symlinks are better, lets wait to hear back from a windows user before reverting anything.

**Comment by janschulz** [comment]
First problem (because I have sphinx in my root env and matplotlib in a special build env -> shinx doesn't see the env matplotlib... :-/):

```bat
c:\data\external\pydata\matplotlib\doc (pr/6663)
[matplotlib_build] λ make.bat html
Running Sphinx v1.3.1

Configuration error:
The configuration file (or one of the modules it imports) called sys.exit()
```

-> Could you change the `sys.exit(msg)` to an import error or a print+`sys.exit(1)`?

After installing sphinx and all the stuff in my env:

```bat
c:\data\external\pydata\matplotlib\doc (pr/6663)
[matplotlib_build] λ make.bat html
Running Sphinx v1.3.1
here
Initializing GitHub plugin
loading pickled environment... not yet created
[autosummary] generating autosummary for: api\afm_api.rst, api\animation_api.rst, api\api_changes.rst, api\artist_api.rst, api\axes_api.rst, api\axis_api.rst, api\backend_bases_api.rst, api\backend_gtkagg_api.rst, api\backend_managers_api.rst, api\backend_pdf_api.rst, ..., users\recipes.rst, users\screenshots.rst, users\shell.rst, users\style_sheets.rst, users\text_intro.rst, users\text_props.rst, users\tight_layout_guide.rst, users\transforms_tutorial.rst, users\usetex.rst, users\whats_new.rst

building [mo]: targets for 0 po files that are out of date
building [html]: targets for 161 source files that are out of date
updating environment: 162 added, 0 changed, 0 removed
reading sources... [  3%] api/axes_api
Exception occurred:
  File "C:\portabel\miniconda\envs\matplotlib_build\lib\site-packages\matplotlib-2.0.0b1+1728.g068cd16.dirty-py3.5-win-amd64.egg\matplotlib\sphinxext\plot_directive.py", line 699, in run
    with io.open(source_file_name, 'r', encoding='utf-8') as fd:
FileNotFoundError: [Errno 2] No such file or directory: 'c:\\data\\external\\pydata\\matplotlib\\doc\\mpl_examples/pylab_examples/xcorr_demo.py'
The full traceback has been saved in C:\Users\jschulz\AppData\Local\Temp\sphinx-err-kbnsomyu.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
```

Is that the expected error or should I dig deeper?

File: doc/users/installing.rst [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+-------------------
+
+The Python shipped from http://www.python.org is compiled with Visual Studio
+2008 for versions before 3.3, Visual Studio 2010 for 3.3 and 3.4, and
+Visual Studio 2015 for 3.5.  Python extensions are recommended to be compiled
+with the same compiler.
+
+Since there is no canonical Windows package manager, the methods for building
+freetype, zlib, and libpng from source code are documented as a build script
+at `matplotlib-winbuild <https://github.com/jbmohler/matplotlib-winbuild>`_.
```
Comment (janschulz) [inline doc/users/installing.rst]: There is also a `build_alllocal.cmd` which has some comments on adding a conda env and then builds it in such an env.

**Comment by NelleV** [comment]
Hi @janschulz 
Can you check in doc that the folder mpl_examples exists?
Thanks,
N

**Comment by janschulz** [comment]
```
c:\data\external\pydata\matplotlib\doc (pr/6663)
[matplotlib_build] λ make.bat html
Running Sphinx v1.3.1
Initializing GitHub plugin
loading pickled environment... not yet created
[autosummary] generating autosummary for: api\afm_api.rst, api\animation_api.rst, api\api_changes.rst, api\artist_api.rst, api\axes_api.rst, api\axis_api.rst, api\backend_bases_api.rst, api\backend_gtkagg_api.rst, api\backend_managers_api.rst, api\backend_pdf_api.rst, ..., users\recipes.rst, users\screenshots.rst, users\shell.rst, users\style_sheets.rst, users\text_intro.rst, users\text_props.rst, users\tight_layout_guide.rst, users\transforms_tutorial.rst, users\usetex.rst, users\whats_new.rst

building [mo]: targets for 0 po files that are out of date
building [html]: targets for 162 source files that are out of date
updating environment: 162 added, 0 changed, 0 removed
reading sources... [  3%] api/axes_api
Exception occurred:
  File "C:\portabel\miniconda\envs\matplotlib_build\lib\site-packages\matplotlib-2.0.0b1+1728.g068cd16.dirty-py3.5-win-amd64.egg\matplotlib\sphinxext\plot_directive.py", line 699, in run
    with io.open(source_file_name, 'r', encoding='utf-8') as fd:
FileNotFoundError: [Errno 2] No such file or directory: 'c:\\data\\external\\pydata\\matplotlib\\doc\\mpl_examples/pylab_examples/xcorr_demo.py'
The full traceback has been saved in C:\Users\jschulz\AppData\Local\Temp\sphinx-err-d2ep60n3.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!

c:\data\external\pydata\matplotlib\doc (pr/6663)
[matplotlib_build] λ ls
Makefile     api/          devel/     make.bat       pyplots/           thirdpartypackages/
README.txt   build/        examples/  matplotlibrc   resources/         users/
_static/     conf.py       faq/       mpl_examples   sphinxext/         utils/
_templates/  contents.rst  glossary/  mpl_toolkits/  style_changes.rst

c:\data\external\pydata\matplotlib\doc (pr/6663)
[matplotlib_build] λ
```

-> I think not.

**Comment by janschulz** [comment]
Having read a bit about it, I think for windows users, the best would be to not check in the `doc\mpl_examples` symlink but add it to gitignore and generate a symlink/copy the files when running the makefile/make.bat file...

**Comment by NelleV** [comment]
@janschulz sounds good. I'll do that.

File: doc/users/installing.rst [hunk 5]
```diff
@@ -0,0 +1,333 @@
+.. The source of this document is INSTALL. During the doc build process,
```
Comment (tacaswell) [inline doc/users/installing.rst]: probably should not commit this file.

**Comment by NelleV** [comment]
I am having trouble with the symlink on windows (more exactly, I am having problems with the make.bat file). 

The reason we need to copy/symlink examples to the matplotlib folder is to be able to link to the documentation example. We could decide to switch to sphinx-gallery to build the gallery, which would allow this feature out of the box. It would also allow to have text displayed in addition to the code when displaying a gallery example, and have a title displayed in the gallery instead of the filename, and generate an ipython notebook in addition to the python file. If I am not mistaken, this is the sort of things sphinx-gallery can do: http://scikit-image.org/docs/dev/auto_examples/color_exposure/plot_ihc_color_separation.html#sphx-glr-auto-examples-color-exposure-plot-ihc-color-separation-py
sphinx-gallery can be added to the sphinxext folder and replace the gen_gallery.py present there. (the code from sphinx-gallery is actually based on matplotlib's gen_gallery.py).

Else I'll resign myself to booting under windows to test this, though I am unsure whether I'll even be able to install matplotlib there (or code).

**Comment by QuLogic** [comment]
Is there a way to do it without a symlink? I think we tried that on another PR and it just wouldn't work for Windows.

Long term, switching to sphinx-gallery is probably a good idea; if you wanted to start that, it might be best to do that before this PR.

**Comment by NelleV** [comment]
On windows, without sphinx-gallery, the folder that is symlinked on linux should be copied at every build.

**Comment by NelleV** [comment]
I am going to close for now, as this is going to take a while to get all the examples sphinx-gallery compatible.

**Pull request closed without merging**